File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/domain/pullrequests/issues/GetAllClosableByInteractor.java
package com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.issues;

import com.duxtinto.intellij.plugin.github.codereviews.domain.Interactor;
import com.duxtinto.intellij.plugin.github.codereviews.domain.issues.IssueEntity;
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.PullRequestDomainContract;
import org.jetbrains.annotations.NotNull;

import javax.annotation.Nonnull;
import java.net.URL;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.google.common.base.Preconditions.checkNotNull;

public class GetAllClosableByInteractor implements Interactor<String, List<IssueEntity>> {

    private final PullRequestDomainContract.DescriptionParser descriptionParser;

    public GetAllClosableByInteractor(PullRequestDomainContract.DescriptionParser descriptionParser) {
        this.descriptionParser = descriptionParser;
    }

    @Override
    public List<IssueEntity> run(@Nonnull String description) {
        checkNotNull(description);

        return transformToEntityList(parseForClosableIssues(description));
    }

    private List<IssueEntity> transformToEntityList(Map<Long, URL> closableIssueMap) {
        return closableIssueMap
                .entrySet()
                .stream()
                .map(entry -> IssueEntity.create(entry.getKey(), entry.getValue()))
                .collect(Collectors.toList());
    }

    @NotNull
    private Map<Long, URL> parseForClosableIssues(@Nonnull String description) {
        return descriptionParser
                .parse(description)
                .closableIssues();
    }
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/domain/pullrequests/issues/GetAllClosableByInteractor.kt
package com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.issues

import com.duxtinto.intellij.plugin.github.codereviews.domain.issues.IssueEntity
import com.duxtinto.intellij.plugin.github.codereviews.domain.issues.IssuesDomainContract
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.PullRequestDomainContract
import javax.inject.Inject

class GetAllClosableByInteractor
    @Inject
    constructor(private val descriptionParser: PullRequestDomainContract.DescriptionParser,
                private val issueFetcher: IssuesDomainContract.Fetcher)
    : ParseIssuesFromStringInteractor {

    override fun run(request: String): List<IssueEntity> {
        return transformToEntityList(parseForClosableIssues(request))
    }

    private fun transformToEntityList(closableIssueMap: Collection<Long>): List<IssueEntity> {
        return closableIssueMap
                .mapNotNull { id ->
                    issueFetcher.fetchIssueById(id)
                }
    }

    private fun parseForClosableIssues(description: String): Collection<Long> {
        return descriptionParser
                .parse(description)
                .closableIssues
    }
}
